File: android/mekuri/src/main/kotlin/studio/anicanon/mekuri/MekuriArrangement.kt
package studio.anicanon.mekuri

import androidx.compose.ui.geometry.Size

/**
 * What the container holds: one page across its whole width, or a spread of two
 * [Spread.pageSize] slots side by side, centred, with the spine between them.
 * Every turn and selection change is judged against it.
 */
internal sealed interface MekuriArrangement {
    data class Single(val pageCount: Int) : MekuriArrangement

    data class Spread(val layout: MekuriSpreadLayout, val pageSize: Size) : MekuriArrangement

    /**
     * The finger's full range across a turn, against which a release is judged:
     * the container for a single page, both slots for a spread.
     */
    fun turnWidth(containerWidth: Float): Float = when (this) {
        is Single -> containerWidth
        is Spread -> this.pageSize.width * 2f
    }

    /**
     * The drag mapping for a turn grabbed at [grabY] in the container. Pages are
     * centred vertically; the grab clamps to the page. [stackTravel] is the spread
     * stack's slide along the drag across the whole turn.
     */
    fun edgeTrack(
        turn: MekuriTurn,
        containerSize: Size,
        grabY: Float,
        liftsFromBottom: Boolean,
        configuration: MekuriConfiguration,
        stackTravel: Float = 0f,
    ): MekuriEdgeTrack {
        val page = when (this) {
            is Single -> containerSize
            is Spread -> this.pageSize
        }
        val y = (grabY - (containerSize.height - page.height) / 2f).coerceIn(0f, page.height)
        return MekuriEdgeTrack(
            turn = turn,
            geometry = MekuriFoldGeometry(page.width, configuration),
            pageHeight = page.height,
            row = if (liftsFromBottom) page.height - y else y,
            reach = this.turnWidth(containerSize.width),
            stackTravel = stackTravel,
        )
    }

    fun turnState(id: Int, turn: MekuriTurn, from: Int): MekuriTurnState = when (this) {
        is Single -> MekuriTurnState.begin(id, turn, from, this.pageCount)
        is Spread -> MekuriTurnState.begin(id, turn, from, this.layout)
    }

    fun transition(from: Int, to: Int): MekuriTransition = when (this) {
        is Single -> MekuriTransition.between(from, to)
        is Spread -> MekuriTransition.between(from, to, this.layout)
    }

    /** Live pages at rest. Single mode fills the trailing slot only. */
    fun slots(page: Int): MekuriSpreadPages = when (this) {
        is Single -> MekuriSpreadPages(
            leading = null,
            trailing = if (page in 0 until this.pageCount) page else null,
        )
        is Spread -> this.layout.pages(this.layout.spreadIndex(page))
    }

    /** Always zero in single mode. */
    fun shiftSpan(page: Int, turn: MekuriTurnState?, direction: MekuriDirection): MekuriShiftSpan =
        when (this) {
            is Single -> MekuriShiftSpan.Zero
            is Spread -> if (turn != null) {
                MekuriSpreadShift.span(turn, this.layout, this.pageSize.width, direction)
            } else {
                val rest = MekuriSpreadShift.atRest(page, this.layout, this.pageSize.width, direction)
                MekuriShiftSpan(rest, rest)
            }
        }

    companion object {
        fun resolve(
            containerSize: Size,
            pageCount: Int,
            spread: MekuriSpread,
            coverStandsAlone: Boolean,
            pageAspectRatio: Float,
        ): MekuriArrangement {
            if (!spread.isDouble(containerSize, pageAspectRatio)) return Single(pageCount)
            return Spread(
                layout = MekuriSpreadLayout(pageCount, coverStandsAlone),
                pageSize = MekuriSpread.pageSize(containerSize, pageAspectRatio),
            )
        }
    }
}

File: ios/Sources/Mekuri/MekuriArrangement.swift
import CoreGraphics

/// What the container holds: one page across its whole width, or a spread
/// of two `pageSize` slots side by side, centred, with the spine between
/// them. Every turn and selection change is judged against it.
enum MekuriArrangement: Equatable {
    case single(pageCount: Int)
    case spread(MekuriSpreadLayout, pageSize: CGSize)

    static func resolve(
        containerSize: CGSize,
        pageCount: Int,
        spread: MekuriSpread,
        coverStandsAlone: Bool,
        pageAspectRatio: CGFloat
    ) -> MekuriArrangement {
        guard spread.isDouble(containerSize: containerSize, pageAspectRatio: pageAspectRatio) else {
            return .single(pageCount: pageCount)
        }
        return .spread(
            MekuriSpreadLayout(pageCount: pageCount, coverStandsAlone: coverStandsAlone),
            pageSize: MekuriSpread.pageSize(fitting: containerSize, pageAspectRatio: pageAspectRatio)
        )
    }

    /// The finger's full range across a turn, against which a release is
    /// judged: the container for a single page, both slots for a spread.
    func turnWidth(containerWidth: CGFloat) -> CGFloat {
        switch self {
        case .single: containerWidth
        case .spread(_, let pageSize): pageSize.width * 2
        }
    }

    /// The drag mapping for a turn grabbed at `grabY` in the container. Pages
    /// are centred vertically; the grab clamps to the page. `stackTravel` is
    /// the spread stack's slide along the drag across the whole turn.
    func edgeTrack(
        turn: MekuriTurn,
        containerSize: CGSize,
        grabY: CGFloat,
        liftsFromBottom: Bool,
        configuration: MekuriConfiguration,
        stackTravel: CGFloat = 0
    ) -> MekuriEdgeTrack {
        let page = switch self {
        case .single: containerSize
        case .spread(_, let pageSize): pageSize
        }
        let y = min(max(grabY - (containerSize.height - page.height) / 2, 0), page.height)
        return MekuriEdgeTrack(
            turn: turn,
            geometry: MekuriFoldGeometry(pageWidth: page.width, configuration: configuration),
            pageHeight: page.height,
            row: liftsFromBottom ? page.height - y : y,
            reach: self.turnWidth(containerWidth: containerSize.width),
            stackTravel: stackTravel
        )
    }

    func turnState(id: Int, turn: MekuriTurn, from page: Int) -> MekuriTurnState {
        switch self {
        case .single(let pageCount):
            MekuriTurnState.begin(id: id, turn: turn, from: page, pageCount: pageCount)
        case .spread(let layout, _):
            MekuriTurnState.begin(id: id, turn: turn, from: page, layout: layout)
        }
    }

    func transition(from: Int, to: Int) -> MekuriTransition {
        switch self {
        case .single:
            MekuriTransition.between(from: from, to: to)
        case .spread(let layout, _):
            MekuriTransition.between(from: from, to: to, layout: layout)
        }
    }

    /// Live pages at rest with `page` selected. Single mode fills the
    /// trailing slot only.
    func slots(showing page: Int) -> (leading: Int?, trailing: Int?) {
        switch self {
        case .single(let pageCount):
            (nil, (0..<pageCount).contains(page) ? page : nil)
        case .spread(let layout, _):
            layout.pages(inSpread: layout.spreadIndex(containing: page))
        }
    }

    /// Shift of the spread stack at rest with `page` selected, or across
    /// `turn`. Always zero in single mode.
    func shiftSpan(showing page: Int, turn: MekuriTurnState?, direction: MekuriDirection) -> MekuriSpreadShift.Span {
        switch self {
        case .single:
            return MekuriSpreadShift.Span.zero
        case .spread(let layout, let pageSize):
            if let turn {
                return MekuriSpreadShift.span(of: turn, layout: layout, pageWidth: pageSize.width, direction: direction)
            }
            let rest = MekuriSpreadShift.atRest(inSpreadContaining: page, layout: layout, pageWidth: pageSize.width, direction: direction)
            return MekuriSpreadShift.Span(start: rest, end: rest)
        }
    }
}
